Add expiration to FunctionOfTime::PiecewisePolynomial

markscheel opened the pull request:
## Proposed changes

The expiration time will be used to tell whether the
FunctionsOfTime in the GlobalCache are up to date.

Interface changes:
 - The one nontrival constructor of PiecewisePolynomial now takes
   an additional value expiration_time.
 - The update function of PiecewisePolynomial now takes an
   expiration_time.
 - PiecewisePolynomial::time_bounds() now returns expiration_time as
   its upper bounds.
 - FunctionOfTime::time_bounds() documentation now states that the
   bounds includes any allowed time-extrapolation.
 - Many tests have been changed to accomodate the interface changes
   in PiecewisePolynomial; this accounts for the vast majority of
   file changes in this commit.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

My current model is to use the upper bound of FunctionOfTime::time_bounds() to determine whether a FunctionOfTime is out of date. Alternatively I could add a new member function FunctionOfTime::expiration_time() expressly for that purpose, but I chose to keep fewer member functions in the interface.

wthrowe commented on src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp:
```diff
@@ -63,13 +65,20 @@ PiecewisePolynomial<MaxDeriv>::func_and_derivs(const double t) const noexcept {
           << deriv_info_at_update_times_.back().time << ".");
   }
+  if (expiration_time <= expiration_time_) {
+    ERROR("expiration_time must be nondecreasing from call to call. "
```
Should there also be a requirement that `time_of_update >= expiration_time_`?  It seems like you would get race conditions if you updated before a time that could have alredy been used.

wthrowe commented on src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp:
```diff
@@ -91,7 +90,7 @@ void PiecewisePolynomial<MaxDeriv>::update(
   }
 
-  func[MaxDeriv] = std::move(updated_max_deriv);
+  func[MaxDeriv] = updated_max_deriv;
```
I'm confused by this one.  It looks like `func` is a `std::array<DataVector, ...>`, which should support move assigning to an element.

markscheel replied:
I had changed this line, and I had simultaneously changed the type of the input argument updated_max_deriv (used to be a DataVector passed by value and I changed it to a `const DataVector &`) because clang-tidy complained about passing by value.  If you want, I can restore the std::move, restore the pass-DataVector-by-value, and add a NOLINT for clang-tidy.

wthrowe replied:
Yeah, I have no idea what clang-tidy is thinking here.  I can reproduce the warnings locally, and it is literally warning that the variable should be made const, and when you do that warning that the variable should not be const.  Just find somewhere to put a NOLINT to shut it up.

kidder commented:
I am confused by what is going on in this PR.  Why is the `expiration_time` being initialized to the `initial_time` in various constructors?  Where is the `ERROR` if I evaluate `func` or `func_and_deriv`, etc. past the `expiration_time`?

markscheel commented:
I did the following:

 - rebased on develop (and did various resulting new clang-tidy fixes in extra fixup commits)
 - pushed fixups to address review of @wthrowe 
 - Fixups: Added ERROR if `func_and_deriv(t)` is called with t past the expiration time (and a test)
 - Fixups: Added a few more ERROR conditions (and tests for them)
 - Fixups: Re-did all the tests so that they don't set expiration_time to initial_time in the constructors, and instead choose some appropriate expiration_time.
 - Fixups: Allow FunctionOfTimeUpdater to update the expiration time if it decides that the FunctionOfTime is ok to not be updated until later.
 - Fixups: Add new option to the 3 DomainCreators that create a PiecewisePolynomial; this option sets the initial expiration time.  And then set this option in the various tests that use these DomainCreators.

Some of my new fixups (particularly the ones that set expiration_time in the tests) redo some of my earlier commits and touch the same code in various files.  Therefore I can reduce the size of this PR a lot if I squash.  Should I squash?

wthrowe commented on src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp:
```diff
@@ -70,7 +70,10 @@ PiecewisePolynomial<MaxDeriv>::func_and_derivs(const double t) const noexcept {
+    // The NOLINT is for performance-unnecessary-value-param
+    // where clang-tidy says to use 'const DataVector& updated_max_deriv'.
+    // However, updated_max_deriv is std::moved out of inside this function.
+    const double time_of_update, DataVector updated_max_deriv,  // NOLINT
```
Use `NOLINT(performance-unnecessary-value-param)` (or the `NOLINTNEXTLINE` version).

wthrowe commented on src/Domain/FunctionsOfTime/PiecewisePolynomial.hpp:
```diff
@@ -60,6 +60,14 @@ class PiecewisePolynomial : public FunctionOfTime {
+  /// keep the current values valid for longer.
+  void reset_expiration_time(double expiration_time) noexcept {
+    expiration_time_ = expiration_time;
+  }
```
This should have a check that the reset doesn't move the time backwards.

wthrowe reviewed (changes requested):
These and the previous fixups can be squashed as far as I am concerned.

kidder commented:
yes, please squash and then I will take a more careful look

markscheel committed:
Clang-tidy

markscheel commented:
Squashed.

markscheel commented:
Oh hold on, I want to move a few changes from one commit to another...

markscheel committed:
Add InitialExpirationDeltaT option to some DomainCreators.

Add this option to CubicScale, UniformTranslation,
and UniformRotationAboutZAxis.

This option effectively sets the initial expiration time.

Also, change the various tests that now need this option.

markscheel committed:
Add expiration time to FunctionsOfTime::PiecewisePolynomial.

The expiration time will be used to tell whether the
FunctionsOfTime in the GlobalCache are up to date.

Interface changes:
 - The one nontrival constructor of PiecewisePolynomial now takes
   an additional value expiration_time.
 - The update function of PiecewisePolynomial now takes an
   expiration_time.
 - There is a new reset_expiration_time function of PiecewisePolynomial
   that is called by FunctionOfTimeUpdater when it decides that
   PiecewisePolynomial should remain valid for longer instead of being
   updated.
 - PiecewisePolynomial::time_bounds() now returns expiration_time as
   its upper bounds.
 - FunctionOfTime::time_bounds() documentation now states that the
   bounds includes any allowed time-extrapolation.
 - Many tests have been changed to accomodate the interface changes
   in PiecewisePolynomial; this accounts for the vast majority of
   files changed in this commit.

markscheel commented:
Ok, attempted to rearrange files in each commit to make things easier to understand.

markscheel commented:
Fixed failing test (problem was a `<=` in an error check that should have been a `<`).

wthrowe commented on src/Domain/FunctionsOfTime/PiecewisePolynomial.cpp:
```diff
@@ -63,13 +70,45 @@ PiecewisePolynomial<MaxDeriv>::func_and_derivs(const double t) const noexcept {
           << deriv_info_at_update_times_.back().time << ".");
   }
+  if (expiration_time <= expiration_time_) {
+    ERROR("expiration_time must be nondecreasing from call to call. "
```
Either "<=" -> "<" or "nondecreasing" -> "increasing".

The current check does not match the one in `reset_expiration_time`.  I don't know if that is intentional.

markscheel commented:
Made change suggested by @wthrowe and squashed (since the change was one character)

markscheel commented:
The failing test was a timeout in tests that I didn't touch, like Swsh. Restarted the tests.

nilsdeppe commented on src/Domain/Creators/TimeDependence/CubicScale.hpp:
```diff
@@ -59,6 +59,12 @@ class CubicScale final : public TimeDependence<MeshDim> {
+  struct UpdateDeltaT {
+    using type = double;
+    static constexpr Options::String help = {
+        "The time interval for updates of the functions of time"};
```
I think you should add that `If the function or expiration time is not updated by the end of the interval the simulation will fail.` because you haven't really supplied any way of saying "It never gets updated." Would this be a good place for using `Auto`?

markscheel replied:
Yes, I agree that UpdateDeltaT is not optimal, partially because there's no way of saying "It never gets updated". Furthermore, I don't like the need to specify UpdateDeltaT here at all, because expiration times should be set by the control system rather than by the DomainCreator constructors.   However, the `FunctionsOfTime::PiecewisePolynomial` needs an expiration time in its constructor, and that constructor  is called by `CubicScale::functions_of_time()`.  Note that `CubicScale::functions_of_time()` is the only place `UpdateDeltaT` is used.

I can see a few possibilities to improve this situation. Either:

 - Rename `UpdateDeltaT` to `InitialExpirationTime` since the option is used only for initializing the `PiecewisePolynomial`. Then use `Options::Auto` to set the automatic value to  `std::numeric_limits<double>::max()` (a.k.a. "it never gets updated"). Rely on the control system to reset the expiration time as needed.
 - Eliminate the option entirely, and set the initial expiration time to `std::numeric_limits<double>::max()` always, expecting that the control system will reset the expiration time as needed.
 
Both of the above  possibilities (as stated) would require me to remove the `next_expiration_time >= expiration_time_` check in `PiecewisePolynomial::reset_expiration_time` (because if the initial expiration time is `std::numeric_limits<double>::max`, the check would always fail).  If we still want the check in `PiecewisePolynomial::reset_expiration_time`, we could do either:
 - Set the initial expiration time to `std::numeric_limits<double>::lowest()` so that the `PiecewisePolynomial` is initially invalid, and then rely on the control system to reset it, including before the first time it gets used.
 - Set the initial expiration time to `initial_time`, so the `PiecewisePolynomial` is initially valid only for its initial time.  Rely on the control system to reset it before it is used for any subsequent times.
 - Change `PiecewisePolynomial::expiration_time_` from a double to a `std::optional<double>`, initially default-constructed, and interpret an empty `expiration_time_` as "never expires". Allow `PiecewisePolynomial::reset_expiration_time` to reset an empty `expiration_time_` to anything.

Thoughts?

markscheel replied:
I currently favor eliminating the option entirely and setting the initial expiration_time to initial_time, because I don't know how the DomainCreator is going to know how to set the initial expiration time.

markscheel replied:
Consensus from SpECTRE meeting Nov 23, 2020: Rename `UpdateDeltaT` -> `InitialExpirationDeltaT`, add `Auto` to this option so that `Auto` makes `expiration_time` become `std::numeric_limits<double>::max()` (i.e. never expires), which is never resettable or updatable.

kidder commented on src/Domain/FunctionsOfTime/PiecewisePolynomial.hpp:
```diff
@@ -56,12 +56,21 @@ class PiecewisePolynomial : public FunctionOfTime {
+  /// Used by `FunctionOfTimeUpdater` to reset the expiration time
+  /// when it decides not to do an update, but instead decides to
+  /// keep the current values valid for longer.
+  void reset_expiration_time(double expiration_time) noexcept;
```
expiration_time - next_expiration_time

kidder commented on src/Domain/FunctionsOfTime/PiecewisePolynomial.hpp:
```diff
@@ -56,12 +56,21 @@ class PiecewisePolynomial : public FunctionOfTime {
-  /// Returns the domain of validity of the function.
+  /// `updated_max_deriv` is a vector of the `MaxDeriv`ths for each component.
+  /// `expiration_time` is the next expiration time.
+  void update(double time_of_update, DataVector updated_max_deriv,
```
expiration_time -> next_expiration_time

kidder reviewed (changes requested):
you can squash

markscheel commented:
Made changes, squashed @kidder 's suggested changes, did not squash last commit that does renaming and Auto.

nilsdeppe reviewed (commented):
I like the `Auto` change and the rename :+1: Thanks for doing that!

I'm happy for you to squash. No need to wait for my approval. I only glanced over the PR to begin with anyway :)

markscheel commented:
Squashed

kidder commented:
@wthrowe please look at
